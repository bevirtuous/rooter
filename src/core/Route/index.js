import { parseUrl } from 'query-string';

const parseOptions = {
  arrayFormat: 'comma',
  parseBooleans: true,
  parseNumbers: true,
};

function Route(options) {
  const { meta = {} } = options;
  const location = options.pathname.trim();
  const id = options.id;

  const splitPath = location.split('#');
  const path = splitPath[0];
  const hash = splitPath[1] || null;
  const { query, url: pathname } = parseUrl(path, parseOptions);

  return {
    id,
    hash,
    location,
    meta,
    pathname,
    query,
  };
}

export default Route;
