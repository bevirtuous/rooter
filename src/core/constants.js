/**
 * PUSH and POP must be uppercase to be compatible
 * with the history package.
 */
export const PUSH = 'PUSH';
export const POP = 'POP';
export const REPLACE = 'REPLACE';
export const RESET = 'RESET';
export const UPDATE = 'UPDATE';

export const ON_PUSH = 'ONPUSH';
export const ON_POP = 'ONPOP';
export const ON_REPLACE = 'ONREPLACE';
export const ON_RESET = 'ONRESET';
export const ON_UPDATE = 'ONUPDATE';
