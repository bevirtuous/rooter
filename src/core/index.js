import router from './Router';
import setQuery from './setQuery';

export const history = {
  back: router.pop,
  current: router.getCurrentRoute,
  push: router.push,
  go: router.push,
  pop: router.pop,
  replace: router.replace,
  reset: router.reset,
  setMeta: router.update,
  setQuery,
};

export {
  EVENT,
  PUSH,
  POP,
  REPLACE,
  RESET,
  UPDATE,
} from './constants';
export { default as emitter } from './emitter';
