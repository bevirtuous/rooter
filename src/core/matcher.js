const pathToRegexp = require('path-to-regexp');

function decodeParam(param) {
  try {
    return decodeURIComponent(param);
  } catch (_) {
    throw new Error();
  }
}

function pathMatch(path, customOptions = {}) {
  const options = {
    sensitive: false,
    strict: false,
    end: false,
    ...customOptions,
  };

  const keys = [];
  const regex = pathToRegexp(path, keys, options);

  return (pathname, params = {}) => {
    const match = regex.exec(pathname);
    const nextParams = { ...params };

    if (!match) {
      return false;
    }

    for (let i = 0; i < keys.length; i += 1) {
      const key = keys[i];
      const param = match[i + 1];

      if (param) {
        nextParams[key.name] = decodeParam(param);

        if (key.repeat) {
          nextParams[key.name] = params[key.name].split(key.delimiter);
        }
      }
    }

    return nextParams;
  };
}

export default pathMatch;
