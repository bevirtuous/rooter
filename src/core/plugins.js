const plugins = [];

export function addPlugin(fn) {
  plugins.push(fn);
}

export function applyPlugins(data) {
  if (plugins.length === 0) {
    return data;
  }

  // Iterate over and call each plugin, passing the initial data.
  // It is important that the accumulated data is merged last
  // so that no plugin can override the work of another.
  return plugins.reduce((acc, plugin) => ({
    ...plugin(data),
    ...acc,
  }), data);
}
