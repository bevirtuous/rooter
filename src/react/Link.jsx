import React from 'react';
import useHistory from './useHistory';

function Link({ to, state = {}, onClick, ...otherProps }) {
  const history = useHistory();

  function handleClick(event) {
    event.preventDefault();

    history.push(to, state);

    onClick && onClick(event);
  }

  return (
    <a {...otherProps} href={to} onClick={handleClick} />
  );
}

export default Link;
