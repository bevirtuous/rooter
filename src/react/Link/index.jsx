import React from 'react';
import { history } from '../../core';

function Link({ children, to, meta = {}, replace = false, onClick, ...other }) {
  function handleClick(event) {
    event.preventDefault();

    if (to) {
      const action = replace ? history.replace : history.push;

      action({ to, meta });

      if (typeof onClick === 'function') {
        onClick(event);
      }
    }
  }

  return (
    <a {...other} href={to} onClick={handleClick}>{children}</a>
  );
}

export default Link;
