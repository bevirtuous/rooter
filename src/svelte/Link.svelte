<script>
  import { history } from '../core';
  
  export let to;
  export let rel = '';
  export let replace = false;
  export let meta = {};

  function handleClick(event) {
    if (to) {
      const action = replace ? history.replace : history.push;
      action({ to, meta });
    }
  }
</script>

<a href={to} {rel} on:click on:click|preventDefault={handleClick}>
  <slot />
</a>
