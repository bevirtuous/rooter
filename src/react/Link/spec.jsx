import React from 'react';
import { shallow } from 'enzyme';
import { history } from '../../core';
import Link from './index';

const spy = jest.spyOn(history, 'push');
const event = {
  preventDefault: jest.fn(),
};

describe('<Link />', () => {
  beforeEach(() => {
    spy.mockClear();
  });

  it('should open link with `to` and `meta`', () => {
    const app = shallow((
      <Link to="/here" meta={{ hi: 123 }}>My Link</Link>
    ));

    app.simulate('click', event);

    expect(app).toMatchSnapshot();
    expect(spy).toHaveBeenCalledWith({
      to: '/here',
      meta: { hi: 123 },
    });
    expect(event.preventDefault).toHaveBeenCalled();
  });

  it('should call the onClick callback', () => {
    const callback = jest.fn();
    const app = shallow((
      <Link to="/here" onClick={callback}>My Link</Link>
    ));

    app.simulate('click', event);

    expect(callback).toHaveBeenCalledWith(event);
  });

  it('should use `history.replace`', () => {
    const replaceSpy = jest.spyOn(history, 'replace');
    const app = shallow((
      <Link to="/here" replace>My Link</Link>
    ));

    app.simulate('click', event);

    expect(replaceSpy).toHaveBeenCalled();
  });

  it('should not use history when `to` is missing', () => {
    const app = shallow((
      // eslint-disable-next-line jsx-a11y/anchor-is-valid
      <Link>My Link</Link>
    ));

    app.simulate('click', event);

    expect(spy).not.toHaveBeenCalled();
  });
});
