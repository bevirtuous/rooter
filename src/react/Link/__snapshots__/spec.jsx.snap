// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> should open link with \`to\` and \`meta\` 1`] = `
<a
  href="/here"
  onClick={[Function]}
>
  My Link
</a>
`;
